Sources/ArgumentParser/Parsable Types/SubcommandsBuilder.swift

```diff
@@ -0,0 +1,65 @@
+//===----------------------------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift Argument Parser open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+/// Build a set of subcommands which can potentially be grouped.
+@resultBuilder
+public struct SubcommandsBuilder {
+  public static func buildExpression(_ single: ParsableCommand.Type) -> [Subcommand] {
+    return [.single(single)]
+  }
+
+  public static func buildExpression(_ group: CommandGroup) -> [Subcommand] {
+    return [.group(group)]
+  }
+
+  public static func buildBlock(_ subcommands: [Subcommand]...) -> [Subcommand] {
+    return subcommands.flatMap { $0 }
+  }
+
+  public static func buildOptional(_ component: [Subcommand]?) -> [Subcommand] {
+    return component ?? []
+  }
+
+  public static func buildEither(first component: [Subcommand]) -> [Subcommand] {
+    return component
+  }
+
+  public static func buildEither(second component: [Subcommand]) -> [Subcommand] {
+    return component
+  }
+
+  public static func buildLimitedAvailability(_ component: [Subcommand]) -> [Subcommand] {
+    return component
+  }
+
+  public static func buildArray(_ components: [[Subcommand]]) -> [Subcommand] {
+    return components.flatMap { $0 }
+  }
+}
+
+/// Describes a single subcommand or a group thereof.
+public enum Subcommand: Sendable {
```

Comment:
This name doesn't quite describe what the type represents, since a group of subcommands is not itself a subcommand. One option is `CommandGroup` – I'm more comfortable with a `CommandGroup` that just stores a single command than a `Subcommand` that stores a bunch of commands. But then we couldn't use that name for the other type, which it's perfect for.

Maybe just making it `Subcommands` is an improvement? I don't think this is a dealbreaker – do you have any other ideas for this name?